src/EditorFeatures/Core.Wpf/InheritanceMargin/InheritanceChainMarginTaggerProvider.cs

```diff
@@ -0,0 +1,132 @@
+                TaggerEventSources.OnViewSpanChanged(ThreadingContext, textViewOpt, TaggerDelay.OnIdle, TaggerDelay.OnIdle),
+                TaggerEventSources.OnDocumentActiveContextChanged(subjectBuffer, TaggerDelay.OnIdle),
+                TaggerEventSources.OnOptionChanged(subjectBuffer, FeatureOnOffOptions.ShowInheritanceMargin, TaggerDelay.OnIdle));
+
+        protected override IEnumerable<SnapshotSpan> GetSpansToTag(ITextView textView, ITextBuffer subjectBuffer)
+        {
+            this.AssertIsForeground();
+
+            var visibleSpan = textView.GetVisibleLinesSpan(subjectBuffer, extraLines: 100);
+            if (visibleSpan == null)
+            {
+                return base.GetSpansToTag(textView, subjectBuffer);
+            }
+
+            return SpecializedCollections.SingletonEnumerable(visibleSpan.Value);
+        }
+
+        protected override async Task ProduceTagsAsync(
+            TaggerContext<InheritanceMarginTag> context,
+            DocumentSnapshotSpan spanToTag,
+            int? caretPosition)
+        {
+            var document = spanToTag.Document;
+            if (document == null)
+            {
+                return;
+            }
+
+            var cancellationToken = context.CancellationToken;
+
+            var options = await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
+            var featureEnabled = options.GetOption(FeatureOnOffOptions.ShowInheritanceMargin);
+            if (!featureEnabled)
+            {
+                return;
+            }
+
+            var inheritanceMarginInfoService = document.WithFrozenPartialSemantics(cancellationToken).GetLanguageService<IInheritanceMarginService>();
+            if (inheritanceMarginInfoService == null)
+            {
+                return;
+            }
+
+            var inheritanceMemberItems = await inheritanceMarginInfoService.GetInheritanceInfoAsync(
+                    document,
+                    spanToTag.SnapshotSpan.Span.ToTextSpan(),
+                    cancellationToken).ConfigureAwait(false);
+
+            if (inheritanceMemberItems.IsEmpty)
+            {
+                return;
+            }
+
+            // One line might have multiple members to show, so group them.
+            // For example:
+            // interface IBar { void Foo1(); void Foo2(); }
+            // class Bar : IBar { void Foo1() { } void Foo2() { } }
+            var lineToMembers = inheritanceMemberItems
+                .GroupBy(item => item.LineNumber)
+                .ToImmutableDictionary(
+                    keySelector: grouping => grouping.Key,
+                    elementSelector: grouping => grouping.SelectAsArray(g => g));
+
+            var snapshot = spanToTag.SnapshotSpan.Snapshot;
+            foreach (var (lineNumber, membersOnTheLine) in lineToMembers)
+            {
+                if (membersOnTheLine.Length > 0)
```

Comment:
this will always be true, right?  you had to have at least one item to haev an item on this line in the first place.

Reply:
Yes, but all the code (include the UI part) would assume there is at least one item,  so I would like to keep this check here in case there is any bug, it would be no margin, not crashing.

Reply:
i guess i'm saying: this is literally the contract of GroupBy.  You can't have a group without elements in it :)  If you want this check, it should be a Contract.ThrowIfFalse/True call, not an actual chec.

Reply:
Sure